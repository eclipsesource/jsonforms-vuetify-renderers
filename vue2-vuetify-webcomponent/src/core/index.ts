export * from './dynamicDefaults';
export * from './i18n';
export * from './json-refs';
export * from './keywords';
export * from './transform';
export * from './types';
export * from './validate';
