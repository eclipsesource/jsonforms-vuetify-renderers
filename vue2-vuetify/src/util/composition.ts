import {
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  JsonFormsSubStates,
  Resolve,
} from '@jsonforms/core';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';
import { useStyles } from '../styles';
import { computed, ComputedRef, inject, ref } from '../vue';
import { Ajv } from 'ajv';

const useControlAppliedOptions = <I extends { control: any }>(input: I) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options)
    )
  );
};

const useComputedLabel = <I extends { control: any }>(
  input: I,
  appliedOptions: ComputedRef<any>
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk
    );
  });
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  I extends { control: any; handleChange: any }
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const isFocused = ref(false);
  const onChange = (value: any) => {
    input.handleChange(input.control.value.path, adaptValue(value));
  };

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  return {
    ...input,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    persistentHint,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <I extends { layout: any }>(input: I) => {
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options)
    )
  );
  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const childLabelForIndex = (index: number) => {
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp)
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms' couldn't be injected. Are you within JSON Forms?"
    );
  }

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};
