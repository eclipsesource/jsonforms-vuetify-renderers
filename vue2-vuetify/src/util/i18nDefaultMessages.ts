export const i18nDefaultMessages = {
  arraylayout: {
    add: 'Add',
    delete: 'Delete',
    moveUp: 'Move Up',
    moveDown: 'Move Down',
    dialogTitle: 'Delete {{ element }}?',
    dialogText: 'The element will be deleted.',
    dialogConfirm: 'Delete',
    dialogCancel: 'Cancel',
  },
};
