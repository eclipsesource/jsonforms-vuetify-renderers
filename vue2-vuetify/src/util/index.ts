export * from './options';
export * from './composition';
export * from './validator';
export * from './datejs';
export * from './i18nDefaultMessages';
