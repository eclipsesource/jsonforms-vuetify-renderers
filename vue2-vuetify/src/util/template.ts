import templateFn from 'lodash/template';
import merge from 'lodash/merge';
import { TemplateExecutor, TemplateOptions } from 'lodash';

const interpolate = /{{([\s\S]+?)}}/g;

export const template = (
  templateString?: string,
  options?: TemplateOptions
): TemplateExecutor => {
  return templateFn(
    templateString,
    merge({ interpolate: interpolate }, options)
  );
};
