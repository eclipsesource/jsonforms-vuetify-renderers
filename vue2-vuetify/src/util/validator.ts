import { Options } from 'ajv';
import { createAjv as createAjvCore } from '@jsonforms/core';
import Ajv from 'ajv';

export const createAjv = (options?: Options): Ajv => {
  const ajv = createAjvCore(options);
  ajv.addFormat('password', (_) => true);
  return ajv;
};
