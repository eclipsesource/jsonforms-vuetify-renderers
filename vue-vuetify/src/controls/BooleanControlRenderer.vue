<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-checkbox
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      :indeterminate="control.data === undefined"
      :model-value="control.data"
      v-bind="vuetifyProps('v-checkbox')"
      @change="onChange"
      @focus="isFocused = true"
      @blur="handleBlur"
    />
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isBooleanControl,
} from '@jsonforms/core';
import { defineComponent } from 'vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue';
import { default as ControlWrapper } from './ControlWrapper.vue';
import {useBlurHandler, useVuetifyControl} from '../util';
import { VCheckbox } from 'vuetify/components';

const controlRenderer = defineComponent({
  name: 'boolean-control-renderer',
  components: {
    VCheckbox,
    ControlWrapper,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    const control = useVuetifyControl(useJsonFormsControl(props), (event) => {
      return event.target.checked || false;
    });
    const { handleBlur } = useBlurHandler(control);
    return {
      ...control,
      handleBlur,
    }
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(1, isBooleanControl),
};
</script>
