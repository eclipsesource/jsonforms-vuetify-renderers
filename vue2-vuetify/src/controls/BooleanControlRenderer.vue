<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-checkbox
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      :indeterminate="control.data === undefined"
      :value="control.data"
      @change="onChange"
      @focus="isFocused = true"
      @blur="isFocused = false"
    />
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isBooleanControl,
} from '@jsonforms/core';
import { defineComponent } from '../vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue2';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { useVuetifyControl } from '../util';
import { VCheckbox } from 'vuetify/lib';

const controlRenderer = defineComponent({
  name: 'boolean-control-renderer',
  components: {
    VCheckbox,
    ControlWrapper,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    return useVuetifyControl(
      useJsonFormsControl(props),
      (newValue) => newValue || false
    );
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(1, isBooleanControl),
};
</script>
