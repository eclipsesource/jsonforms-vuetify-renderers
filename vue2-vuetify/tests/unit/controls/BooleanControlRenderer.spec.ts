import { clearAllIds } from '@jsonforms/core';
import { Wrapper } from '@vue/test-utils';
import BooleanControlRenderer, {
  entry as booleanControlRendererEntry,
} from '../../../src/controls/BooleanControlRenderer.vue';
import { mountJsonForms } from '../util';

describe('BooleanControlRenderer.vue', () => {
  const renderers = [booleanControlRendererEntry];

  const data = true;
  const schema = {
    type: 'boolean',
    title: 'My Boolean',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'boolean placeholder',
    },
  };
  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child BooleanControlRenderer exists', () => {
    expect(wrapper.getComponent(BooleanControlRenderer));
  });

  it('renders a checkbox', () => {
    expect(wrapper.find('input[type="checkbox"]').exists()).toBe(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('My Boolean');
  });

  it('emits a data change', async () => {
    const input = wrapper.find('input[type="checkbox"]');
    await input.trigger('click');
    expect(wrapper.vm.$data.data).toEqual(false);
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="checkbox"]');
    await input.trigger('focus');
    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('boolean placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
