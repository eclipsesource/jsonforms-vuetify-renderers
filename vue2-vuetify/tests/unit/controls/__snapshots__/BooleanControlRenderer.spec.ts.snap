// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanControlRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My Boolean"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-input--selection-controls__input">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-checkbox-marked theme--light primary--text"
              >
              </i>
              <input aria-checked="true"
                     id="#-input"
                     role="checkbox"
                     type="checkbox"
                     placeholder="boolean placeholder"
                     value="true"
              >
              <div class="v-input--selection-controls__ripple primary--text">
              </div>
            </div>
            <label for="#-input"
                   class="v-label theme--light"
                   style="left: 0px; position: relative;"
            >
              My Boolean
            </label>
          </div>
          <div class="v-messages theme--light primary--text">
            <transition-group-stub tag="div"
                                   name="message-transition"
                                   class="v-messages__wrapper"
            >
            </transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
